[JENKINS-21670] Add support for anonymizing support bundle output

jvz opened the pull request:
See [JENKINS-21670](https://issues.jenkins-ci.org/browse/JENKINS-21670).

This adds a new `ContentFilter` extension point for filtering text data written to the support bundle along with supporting code to enable the feature.

This PR is based on #142 which had too many comments on it to reliably load the page. See previous discussions there.

@reviewbybees @cloudbees/team-pre

dwnusbaum commented on src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java:
```diff
@@ -260,136 +264,158 @@ public static void setLogLevel(Level level) {
+                        OutputStreamSelector selector = new OutputStreamSelector(() -> binaryOut, () -> textOut);
+                        IgnoreCloseOutputStream out = new IgnoreCloseOutputStream(selector);
+                        content.writeTo(out);
+                        out.flush();
```
What if there was a `Content` with fewer bytes than the selector uses to make its selection? Do we need an `OutputStreamSelector#forceSelection` method that is called before `out.flush`?

dwnusbaum replied:
Right now it looks like in `OutputStreamSelector#probeContents` that we would delay the selection (`head.hasRemaining()` is true) and not write anything unless `#close` was called.

jvz replied:
Excellent point. I'll update this.

dwnusbaum replied:
To be fair, seems unlikely in practice except for maybe `FileContent` with a tiny file, but could be bad if we later changed `DEFAULT_PROBE_SIZE`.

jvz replied:
I think this could be addressed by adding a call to `OutputStreamSelector::chooseStream` inside `flush`.

dwnusbaum commented on src/main/java/com/cloudbees/jenkins/support/api/PrintedContent.java:
```diff
@@ -42,12 +47,29 @@ public PrintedContent(String name) {
+        if (os instanceof WrapperOutputStream) {
+            OutputStream out = ((WrapperOutputStream) os).getRecursivelyUnderlyingStream();
+            if (out instanceof FilteredOutputStream) {
+                FilteredOutputStream filteredStream = (FilteredOutputStream) out;
```
This is broken again. Since a new `OutputStreamSelector` is created for each `Content`, when this method is called `OutputStreamSelector#getUnderlyingStream` returns null because it has not seen any bytes. We'd want to force the text stream in this case, but I'm not sure what the cleanest approach would be.

jvz replied:
Yeah, I'm not exactly sure what the best approach is here for an optimization. A `ThreadLocal` could be a nice hack if needed.

jvz replied:
If the default behavior of the class is to use the text stream rather than the binary stream for 0 bytes, then this would be consistent and less of a hack. I'll try it out and see what it looks like.

dwnusbaum commented on src/main/java/com/cloudbees/jenkins/support/filter/ContentMappings.java:
```diff
@@ -0,0 +1,193 @@
+        stopWords = proxy.stopWords == null ? getDefaultStopWords() : proxy.stopWords;
+        mappings = proxy.stopWords == null
+                ? new ConcurrentHashMap<>()
+                : proxy.mappings.stream().collect(toConcurrentMap(ContentMapping::getOriginal, Function.identity(), (a, b) -> {throw new IllegalArgumentException();}, ConcurrentSkipListMap::new));
```
The mappings don't use the sort order defined in ContentMapping::COMPARATOR, they sort based on the order of the keys, so I am still seeing the problem where if I have a view named `Test` mapped to `Foo` and a view named `Prefix_Test` mapped to `Bar`, I see things like `Prefix_view_Foo` in the output. I was able to fix it with the following patch (probably worth adding a test though and maybe removing the comparable implementation from ContentMapping if it is unused):

    diff --git a/src/main/java/com/cloudbees/jenkins/support/filter/ContentMappings.java b/src/main/java/com/cloudbees/jenkins/support/filter/ContentMappings.java
    index 22c0684..ae9c488 100644
    --- a/src/main/java/com/cloudbees/jenkins/support/filter/ContentMappings.java
    +++ b/src/main/java/com/cloudbees/jenkins/support/filter/ContentMappings.java
    @@ -37,13 +37,13 @@ import javax.annotation.Nonnull;
     import java.io.IOException;
     import java.util.Arrays;
     import java.util.Collections;
    +import java.util.Comparator;
     import java.util.HashSet;
     import java.util.Iterator;
     import java.util.Locale;
     import java.util.Map;
     import java.util.Set;
     import java.util.Spliterator;
    -import java.util.concurrent.ConcurrentHashMap;
     import java.util.concurrent.ConcurrentSkipListMap;
     import java.util.function.Consumer;
     import java.util.function.Function;
    @@ -61,6 +61,10 @@ import static java.util.stream.Collectors.toMap;
     @Restricted(NoExternalUse.class)
     public class ContentMappings extends ManagementLink implements Saveable, Iterable<ContentMapping> {
     
    +    private static final Comparator<String> BY_LENGTH = Comparator.comparingLong(String::length);
    +    private static final Comparator<String> BY_NAME = Comparator.comparing(Function.identity());
    +    private static final Comparator<String> COMPARATOR = BY_LENGTH.reversed().thenComparing(BY_NAME);
    +
         /**
          * @return the singleton instance
          */
    @@ -87,8 +91,8 @@ public class ContentMappings extends ManagementLink implements Saveable, Iterabl
         private ContentMappings(@Nonnull XmlProxy proxy) {
             stopWords = proxy.stopWords == null ? getDefaultStopWords() : proxy.stopWords;
             mappings = proxy.stopWords == null
    -                ? new ConcurrentHashMap<>()
    -                : proxy.mappings.stream().collect(toConcurrentMap(ContentMapping::getOriginal, Function.identity(), (a, b) -> {throw new IllegalArgumentException();}, ConcurrentSkipListMap::new));
    +                ? new ConcurrentSkipListMap<>(COMPARATOR)
    +                : proxy.mappings.stream().collect(toConcurrentMap(ContentMapping::getOriginal, Function.identity(), (a, b) -> {throw new IllegalArgumentException();}, () -> new ConcurrentSkipListMap(COMPARATOR)));
         }

jvz replied:
Good catch. I've added a test for this as well by verifying the ordering of mappings created when creating some freestyle projects.

dwnusbaum commented on src/main/java/com/cloudbees/jenkins/support/filter/NameProvider.java:
```diff
@@ -0,0 +1,134 @@
+     */
+    public static final @Extension NameProvider ITEMS = new NameProvider(
+            () -> Jenkins.get()
+                    .getItems().stream()
```
`getItems` only returns the top level of items. If you have a job in a folder then the folder will be included, but not the job. You should use `allItems` instead to do a recursive lookup.

dwnusbaum commented on src/main/java/com/cloudbees/jenkins/support/filter/NameProvider.java:
```diff
@@ -0,0 +1,134 @@
+    public static final @Extension NameProvider ITEMS = new NameProvider(
+            () -> Jenkins.get()
+                    .getItems().stream()
+                    .flatMap(item -> Arrays.stream(item.getFullName().split("/")))
```
I think you should be able to use `getName` instead of `getFullName` once you are using `allItems` since `getFullName` just includes the names of all parents up through the root, and you will include them anyway when using `allIItems`.

However, I think you also need to add `item.getDisplayName`, which might appear in logs and can be distinct from the item's name.

jvz replied:
Sounds good to me. Do you think `getDisplayName` should be added for any other categories?

dwnusbaum replied:
I don't know if the other ones can be modified to be distinct from `getName`, but subclasses could definitely support it, so it is probably safest to add it for everything that has a `getDisplayName` method.

dwnusbaum commented on src/main/java/com/cloudbees/jenkins/support/filter/SensitiveContentFilter.java:
```diff
@@ -0,0 +1,71 @@
+        Set<String> stopWords = mappings.getStopWords();
+        for (NameProvider provider : NameProvider.all()) {
+            provider.names()
+                    .filter(name -> !stopWords.contains(name.toLowerCase(Locale.ENGLISH)))
```
After the latest commit, tests are failing because `Computer#getName` sometimes returns "", which makes it into a mapping here and causes all kinds of strangeness.

I would add something like the following to fix the tests:

    diff --git a/src/main/java/com/cloudbees/jenkins/support/filter/SensitiveContentFilter.java b/src/main/java/com/cloudbees/jenkins/support/filter/SensitiveContentFilter.java
    index e0f3660..42893a3 100644
    --- a/src/main/java/com/cloudbees/jenkins/support/filter/SensitiveContentFilter.java
    +++ b/src/main/java/com/cloudbees/jenkins/support/filter/SensitiveContentFilter.java
    @@ -32,6 +32,7 @@ import org.kohsuke.accmod.restrictions.NoExternalUse;
     import javax.annotation.Nonnull;
     import java.util.Locale;
     import java.util.Set;
    +import org.apache.commons.lang.StringUtils;
     
     /**
      * Filters contents based on names provided by all {@linkplain NameProvider known sources}.
    @@ -63,6 +64,7 @@ public class SensitiveContentFilter implements ContentFilter {
             Set<String> stopWords = mappings.getStopWords();
             for (NameProvider provider : NameProvider.all()) {
                 provider.names()
    +                    .filter(name -> !StringUtils.isBlank(name))
                         .filter(name -> !stopWords.contains(name.toLowerCase(Locale.ENGLISH)))
                         .forEach(name -> mappings.getMappingOrCreate(name, original -> ContentMapping.of(original, provider.generateFake())));
             }

jvz replied:
Whoops! Fixed. There's even a `StringUtils.isNotBlank(name)` which reads well.

jvz commented:
The build pipeline has an error completely unrelated to code changes, and I can't restart the build (even when logged in to ci.jenkins.io).

jvz commented:
@stephenc @christ66 can we get a review of this PR?

christ66 commented on src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java:
```diff
@@ -428,29 +457,21 @@ public static void threadDumpStartup() throws Exception {
-                    } finally {
-                        org.apache.commons.io.IOUtils.closeQuietly(ps);
-                        org.apache.commons.io.IOUtils.closeQuietly(fileOutputStream);
+                        Thread.currentThread().interrupt();
```
Is it necessary to do an interrupt at the end of the finally block? This would just generate a thread dump once at the start and then immediately kill the thread. It would be best if exceptions are thrown to display the message, and then do a `return;`.

christ66 replied:
I think the second sentence is the key part. This code is designed so at the start of Jenkins it will generate a thread dump ever `X` seconds. The `Thread.currentThread().interrupt();` will kill the thread after the first thread dump. Defeats the purpose of this feature. If you merge all the catch blocks into one, and print the stack trace, then do the `Thread.currentThread().interrupt();` then it will only stop generating thread dumps in the following cases:
* Once Jenkins is online.
* It was unable to save/load/generate the thread dump or files.
Looking at this code I wrote I realize it's partially (....ok completely) wrong. I didn't realize, but every time I generate a thread dump I close and re-open the resource. It's probably best to move the try/catch/finally block outside of the `while` loop. Then the finally block with the interrupt call would be safe.

jvz commented:
It’s generally best practice to reinterrupt when catching that. I’ll see if
I can find some better details.

On Thu, Jun 14, 2018 at 19:29, Steven Christou <notifications@github.com>
wrote:

> *@christ66* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java
> <https://github.com/jenkinsci/support-core-plugin/pull/144#discussion_r195608692>
> :
>
> >                          e.printStackTrace();
>                      } catch (InterruptedException e) {
>                          e.printStackTrace();
> -                    } finally {
> -                        org.apache.commons.io.IOUtils.closeQuietly(ps);
> -                        org.apache.commons.io.IOUtils.closeQuietly(fileOutputStream);
> +                        Thread.currentThread().interrupt();
>
> Is it necessary to do an interrupt at the end of the finally block? This
> would just generate a thread dump once at the start and then immediately
> kill the thread. It would be best if exceptions are thrown to display the
> message, and then do a return;.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/support-core-plugin/pull/144#pullrequestreview-129013369>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAwS635z3X3psbmLZrETYU7hY4b4KyCgks5t8v_8gaJpZM4UB7ZS>
> .
>
-- 
Matt Sicker
Software Engineer, CloudBees

christ66 commented on src/main/java/com/cloudbees/jenkins/support/filter/FilteredOutputStream.java:
```diff
@@ -0,0 +1,222 @@
+@Restricted(NoExternalUse.class)
+public class FilteredOutputStream extends FilterOutputStream {
+
+    static final Pattern EOL = Pattern.compile("\r?\n");
```
Would it be possible to merge the variables from `FilteredWriter` which also has `EOL` and `DEFAULT_DECODER_CAPACITY`?

jvz commented:
@christ66 addressed your comments. As for the synchronized methods, most implementations of `java.io` classes are fully synchronized, so I don't see any issue there. I made sure to annotate protected data, and due to some strange use cases, findbugs helped me refactor things to be safer anyways. :)

dwnusbaum commented:
@christ66 I'm going to go ahead and merge and release, but please feel free to comment if you have any unaddressed concerns and we can handle them in a follow-up PR. Thanks!
